Add inspector monitor

jhen0409 opened the pull request:
Add inspector monitor and create a select tag for switch monitors:

![2016-04-04 11 33 44](https://cloud.githubusercontent.com/assets/3001525/14253344/b5b6f666-fabd-11e5-897c-79ce29b3fdcb.png)

But if store don't has all of state, Diff and State tab will be empty.

zalmoxisus commented:
That's great! I was thinking of many implementations of supporting multiple monitors, but just adding a selector would be probably the best.

Wouldn't be good to have the bottom slider monitor always visible regardless of each monitor we select? Or is there an issue of space or is it useless in case of the inspector monitor?

> But if store don't has all of state, Diff and State tab will be empty.

It will be empty even when we get new data? I thought I solved it in https://github.com/alexkuz/redux-devtools-inspector/pull/22

jhen0409 commented:
> Or is there an issue of space or is it useless in case of the inspector monitor?

Looks no issues, I'll update it. :)

> It will be empty even when we get next data? I thought I solved it in alexkuz/redux-devtools-inspector#22

It seems I restarted remotedev-app, and the application is keep running, the remotedev store will get the following actions: (not have `@@INIT`)

```
increment
increment
decrement
```

I traced the Inspector's log:

```
computedStates: [Object], currentActionId: 15, actions: Object {15: Object}, fromState: undefined, toState: undefined
computedStates: [Object, Object], currentActionId: 16, actions: Object {15: Object, 16: Object}, fromState: undefined, toState: undefined
computedStates: [Object, Object, Object], currentActionId: 17, actions: Object {15: Object, 16: Object, 17: Object}, fromState: undefined, toState: undefined
```

It seems no way to get the correct computedState with currentActionId.

zalmoxisus commented:
Thanks, @jhen0409.

Then it would be a problem there when filtering actions as well. I'll to find a way to get the correct computedState with currentActionId and will make a pr for `inspector monitor` if, of course, you don't have a solution and won't do it first :)

zalmoxisus commented:
Thanks, @jhen0409.

Then it would be a problem there when filtering actions as well. I'll to find a way to get the correct computedState with currentActionId and will make a pr for `inspector monitor` if, of course, you don't have a solution and won't do it first :)

jhen0409 commented:
I think let `remotedev-app` keep has all states will be better. @zalmoxisus what do you think?

1. Change `remote-redux-devtools` filters, let filter actions still send to `remotedev-server`, but it's hidden on `remotedev-app`.
2. A new behavior when `remotedev-app` not have `@@INIT` state, requires synchronization.

zalmoxisus commented:
1. It will make no difference for `redux-devtools-inspector` whether we don't send those actions at all or strip them via filter-action monitor. The same issue is for Redux Devtools Extension and with vanilla DevTools with filter-action monitor.
2. We had this behaviour before, but there were some issues with multiple instances and multiple monitors. I'll look into it again when we support `maxAge`, otherwise it could freeze the app for some minutes.

jhen0409 commented:
@zalmoxisus I removed the comment, sorry. :\

> The same issue is for Redux Devtools Extension and with vanilla DevTools with filter-action monitor.

Yes, It seems that we cannot enable filters in inspector & diff-monitor.

zalmoxisus commented:
@jhen0409, I think it should be an easy fix. The state index should correspond to the action index regardless of the data consistence.

jhen0409 commented:
Due to the `log-monitor` still need original array, I added refreshState function for `liftedStore`, used when switching `log-monitor` and `inspector`.

But it looks `slider-monitor` will not work when I switched to inspector.

zalmoxisus commented:
Better to fix this on the `redux-devtools-inspector` side than adding workarounds here. Also you should take into consideration that in future we would like the app to remember the last selected monitor, not to select it every time.

zalmoxisus commented:
A lot of changes from here would be a breaking change for Redux DevTools Extension, so it should be released as `v0.3.0`.

jhen0409 commented:
@zalmoxisus I made a PR for `redux-devtools-inspector` and removed a last commit.

zalmoxisus commented:
Oh, you are fast :-), I was just looking into it! Thanks! I will add my thoughts there then.

zalmoxisus commented on src/app/utils/localStorage.js:
```diff
@@ -9,6 +9,11 @@ export function getSettings() {
 }
 
+export function getSelectMonitor() {
+  if (localStorage.isChromeStorage) return null;
```
What `localStorage.isChromeStorage` stands for? As far as I see it is always undefined, not?

zalmoxisus commented:
Thanks a lot, @jhen0409!

zalmoxisus commented:
Released in `v0.3.0`. I also [moved everything related to monitors into `DevTools.js`](https://github.com/zalmoxisus/remotedev-app/commit/65cead2ab3cea4373dd03477edfbd49b12e2b917)  to be able to reuse it in Redux Devtools Extension.
